File: MW_Test/MwSesnsorConnectionViewController.m
//
//  MwSesnsorConnectionViewController.m
//  MW_Test
//
//  Created by Rio Shao on 12/6/23.
//

#import "MwSesnsorConnectionViewController.h"

@interface MwSesnsorConnectionViewController ()

@end

@implementation MwSesnsorConnectionViewController

- (void) viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    _mwDevices = [[MWSensor alloc] init];
    _mwDevices = [_mwDevices getMwSensonrs];
    
    _LeftProgressBar.progress = 0;
    _RightProgressBar.progress = 0;
    
    [self ConnectSavedDevices];
}

- (void)viewDidLoad
{
    [super viewDidLoad];
}


- (void) ConnectSavedDevices
{
    [self UpdateSensorConnectButtonFetching: SensorTypeLeft];
    [self UpdateSensorConnectButtonFetching: SensorTypeRight];
    
    [_mwDevices fetchSavedMetawearWithCompletion:
     ^(SensorType type){dispatch_async(dispatch_get_main_queue(), ^{
        [self UpdateSensorConnectButtonConnected:type];
        [self->_mwDevices startMeteringWithType:type];
        [self->_mwDevices readBatteryLevel];
        [self UpdateBatteryLevel];
    });}];
}

- (void) UpdateSensorConnectButtonFetching: (SensorType) type
{
    UIButton *targetButton = [[UIButton alloc] init];
    if(type == SensorTypeLeft)
    {
        targetButton = _LeftSensorConnectButton;
    }
    else if(type == SensorTypeRight)
    {
        targetButton = _RightSensorConnectButton;
    }
    [targetButton setTitleColor:UIColor.blueColor forState:UIControlStateNormal];
    [targetButton setTitle:@"Fetching" forState:UIControlStateNormal];
    targetButton.enabled = YES;
}

- (void) UpdateSensorConnectButtonConnect: (SensorType) type
{
    UIButton *targetButton = [[UIButton alloc] init];
    if(type == SensorTypeLeft)
    {
        targetButton = _LeftSensorConnectButton;
    }
    else if(type == SensorTypeRight)
    {
        targetButton = _RightSensorConnectButton;
    }
    [targetButton setTitleColor:UIColor.blueColor forState:UIControlStateNormal];
    [targetButton setTitle:@"Connect" forState:UIControlStateNormal];
    targetButton.enabled = YES;
}

- (void) UpdateSensorConnectButtonConnecting: (SensorType) type
{
    UIButton *targetButton = [[UIButton alloc] init];
    if(type == SensorTypeLeft)
    {
        targetButton = _LeftSensorConnectButton;
    }
    else if(type == SensorTypeRight)
    {
        targetButton = _RightSensorConnectButton;
    }
    [targetButton setTitleColor:UIColor.grayColor forState:UIControlStateNormal];
    [targetButton setTitle:@"Connecting" forState:UIControlStateNormal];
    targetButton.enabled = NO;
}

- (void) UpdateSensorConnectButtonConnected: (SensorType) type
{
    UIButton *targetButton = [[UIButton alloc] init];
    if(type == SensorTypeLeft)
    {
        targetButton = _LeftSensorConnectButton;
    }
    else if(type == SensorTypeRight)
    {
        targetButton = _RightSensorConnectButton;
    }
    [targetButton setTitleColor:UIColor.greenColor forState:UIControlStateNormal];
    [targetButton setTitle:@"Connected" forState:UIControlStateNormal];
    targetButton.enabled = NO;
}


- (void) UpdateBatteryLevel
{
    // tries to read battery level from sensors for every 1 second,
    // until we sucessfully get the readings
    dispatch_async(dispatch_get_main_queue(), ^{
            [NSTimer scheduledTimerWithTimeInterval:0.2
                                            repeats:YES
                                              block:^(NSTimer *timer)
             {
                float leftLevel = (float)[self->_mwDevices getBatteryLevelWithType:SensorTypeLeft]/100.0;
                float rightLevel = (float)[self->_mwDevices getBatteryLevelWithType:SensorTypeRight]/100.0;
                
                self->_LeftProgressBar.progress = leftLevel;
                self->_RightProgressBar.progress = rightLevel;
                
                if(leftLevel != 0 && rightLevel != 0)
                {
                    [timer invalidate];
                }
            }];
        });
}

- (IBAction)RightSensorConnectButtonPressed:(id)sender {
    int type = SensorTypeRight;
    [self UpdateSensorConnectButtonConnecting:type];
    [self pickUpSensor:type];
}

- (IBAction)LeftSensorConnectButtonPressed:(id)sender {
    int type = SensorTypeLeft;
    [self UpdateSensorConnectButtonConnecting:type];
    [self pickUpSensor:type];
}

- (void) pickUpSensor: (SensorType) type
{
    [_mwDevices resetConnectionWithType:type];
    [_mwDevices scanMetawearWithType:type completion:^{
        dispatch_async(dispatch_get_main_queue(), ^{
            [self UpdateSensorConnectButtonConnected:type];
            [self->_mwDevices startMeteringWithType:type];
            [self->_mwDevices readBatteryLevel];
            [self UpdateBatteryLevel];
        });
    }];
    
    timerCount = 10;
    [NSTimer scheduledTimerWithTimeInterval:1.0
                                    repeats:YES
                                      block:^(NSTimer *timer){
        self->timerCount = self->timerCount - 1;
        //printf("\(type.text) scan: \(counter)");
        if ([self->_mwDevices getMwConnectedWithType:type])
        {
            [self->_mwDevices stopScan];
            [timer invalidate];
        }
        else if (self->timerCount == 0)
        {
            [self->_mwDevices stopScan];
            [self UpdateSensorConnectButtonConnect:type];
            [timer invalidate];
        }
    }];
}

- (IBAction)ExitButtonPressed:(id)sender {
    [_mwDevices savePreferences];
}

@end

File: MW_Test/MwSesnsorConnectionViewController.swift
import UIKit

@objc
class MwSesnsorConnectionViewController: UIViewController, BatteryLevelDelegate {
    
    
    var timerCount: Int = 0
    
    @IBOutlet var LeftProgressBar: UIProgressView!
    @IBOutlet var RightProgressBar: UIProgressView!
    @IBOutlet var RightSensorConnectButton: UIButton!
    @IBOutlet var LeftSensorConnectButton: UIButton!
    @IBOutlet var ExitButton: UIButton!
    
    let mwDevices = AssessmentSettings.sharedManager.mwSensors
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        mwDevices.batteryLevelDelegate = self
        
        LeftProgressBar.progress = 0
        RightProgressBar.progress = 0
        
        connectSavedDevices()
    }
    
    func connectSavedDevices() {
        updateSensorConnectButtonConnecting(type:.left)
        updateSensorConnectButtonConnecting(type:.right)
        
        mwDevices.updateStatus()
        mwDevices.fetchSavedMetawear { [weak self] type in
            DispatchQueue.main.async {
                guard let self = self else { return }
                self.updateSensorConnectButtonConnected(type: type)
                self.mwDevices.readBatteryLevel()
            }
        }
        
        activateConnectionTimerCountdown(type: .left)
        activateConnectionTimerCountdown(type: .right)
        
    }
    
    func batteryLevelUpdate(type: MWSensor.SensorType, batteryLevel: Float) {
        var targetProgressBar : UIProgressView
        switch type {
        case .left:
            targetProgressBar = LeftProgressBar
        case .right:
            targetProgressBar = RightProgressBar
        }
        
        DispatchQueue.main.async {
            targetProgressBar.progress = batteryLevel
        }
    }
    
    
    func updateSensorConnectButtonConnect(type: MWSensor.SensorType) {
        let targetButton: UIButton
        switch type {
        case .left:
            targetButton = LeftSensorConnectButton
        case .right:
            targetButton = RightSensorConnectButton
        }
        targetButton.setTitleColor(.blue, for: .normal)
        targetButton.setTitle("Connect", for: .normal)
        targetButton.isEnabled = true
    }
    
    func updateSensorConnectButtonConnecting(type: MWSensor.SensorType) {
        let targetButton: UIButton
        switch type {
        case .left:
            targetButton = LeftSensorConnectButton
        case .right:
            targetButton = RightSensorConnectButton
        }
        targetButton.setTitleColor(.gray, for: .normal)
        targetButton.setTitle("Connecting", for: .normal)
        targetButton.isEnabled = false
    }
    
    func updateSensorConnectButtonConnected(type: MWSensor.SensorType) {
        let targetButton: UIButton
        switch type {
        case .left:
            targetButton = LeftSensorConnectButton
        case .right:
            targetButton = RightSensorConnectButton
        }
        targetButton.setTitleColor(.green, for: .normal)
        targetButton.setTitle("Connected", for: .normal)
        targetButton.isEnabled = true
    }
    
    
    @IBAction func RightSensorConnectButtonPressed(_ sender: Any) {
        updateSensorConnectButtonConnecting(type:.right)
        pickUpSensor(type:.right)
    }
    
    @IBAction func LeftSensorConnectButtonPressed(_ sender: Any) {
        updateSensorConnectButtonConnecting(type:.left)
        pickUpSensor(type:.left)
    }
    
    @IBAction func ExitButtonPressed(_ sender: Any) {
        mwDevices.savePreferences()
    }
    
    func pickUpSensor(type: MWSensor.SensorType) {
        mwDevices.resetConnection(type: type)
        mwDevices.scanMetawear(type: type) { [weak self] in
            DispatchQueue.main.async {
                guard let self = self else { return }
                self.updateSensorConnectButtonConnected(type: type)
                self.mwDevices.readBatteryLevel()
                //self.updateBatteryLevel()
            }
        }
        activateConnectionTimerCountdown(type: type)
        
    }
    
    func activateConnectionTimerCountdown (type: MWSensor.SensorType)
    {
        var counter = 10 // no of seconds to wait for scan
        Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { timer in
            counter = counter - 1
            print("\(type.text) scan: \(counter)")
            if self.mwDevices.MWConnected[type] == true || counter == 0
            {
                timer.invalidate()
                self.mwDevices.stopScan()
            }
            if self.mwDevices.MWConnected[type] == false && counter == 0
            {
                DispatchQueue.main.async {
                    self.updateSensorConnectButtonConnect(type:type)
                }
            }
        }
    }
}
